pkgs/development/compilers/llvm/default.nix

```diff
@@ -0,0 +1,101 @@
+{
+  lib,
+  recurseIntoAttrs,
+  callPackages,
+  stdenvAdapters,
+  buildPackages,
+  targetPackages,
+  stdenv,
+  gcc12Stdenv,
+  pkgs,
+  # This is the default binutils, but with *this* version of LLD rather
+  # than the default LLVM version's, if LLD is the choice. We use these for
+  # the `useLLVM` bootstrapping below.
+  bootBintoolsNoLibc ? if stdenv.targetPlatform.linker == "lld" then null else pkgs.bintoolsNoLibc,
+  bootBintools ? if stdenv.targetPlatform.linker == "lld" then null else pkgs.bintools,
+}:
+let
+  versions = [
+    {
+      officialRelease = {
+        version = "13.0.1";
+        sha256 = "06dv6h5dmvzdxbif2s8njki6h32796v368dyb5945x8gjj72xh7k";
+      };
+    }
+    {
+      officialRelease = {
+        version = "14.0.6";
+        sha256 = "sha256-vffu4HilvYwtzwgq+NlS26m65DGbp6OSSne2aje1yJE=";
+      };
+    }
+    {
+      officialRelease = {
+        version = "15.0.7";
+        sha256 = "sha256-wjuZQyXQ/jsmvy6y1aksCcEDXGBjuhpgngF3XQJ/T4s=";
+      };
+    }
+    {
+      officialRelease = {
+        version = "16.0.6";
+        sha256 = "sha256-fspqSReX+VD+Nl/Cfq+tDcdPtnQPV1IRopNDfd5VtUs=";
+      };
+    }
+    {
+      officialRelease = {
+        version = "17.0.6";
+        sha256 = "sha256-8MEDLLhocshmxoEBRSKlJ/GzJ8nfuzQ8qn0X/vLA+ag=";
+      };
+    }
+    {
+      officialRelease = {
+        version = "18.1.8";
+        sha256 = "sha256-iiZKMRo/WxJaBXct9GdAcAT3cz9d9pnAcO1mmR6oPNE=";
+      };
+    }
+    {
+      gitRelease = {
+        version = "19.0.0-git";
+        rev = "9b9405621bcc55b74d2177c960c21f62cc95e6fd";
+        rev-version = "19.0.0-unstable-2024-06-30";
+        sha256 = "sha256-Tlk+caav7e7H6bha8YQyOl+x2iNk9iH7xKpHQkWQyJ4=";
+      };
+    }
+  ];
```

Comment:
There are a couple of ways to implement this and I'm not deadset on anything specific, just adding suggestions.

Suggestion, untested! - feel free to ignore: condensed but possibly a bit opaque:
```nix
versions = {
  "13.0.1".officialRelease.sha256 = "06dv6h5dmvzdxbif2s8njki6h32796v368dyb5945x8gjj72xh7k";
  "14.0.6".officialRelease.sha256 = "sha256-vffu4HilvYwtzwgq+NlS26m65DGbp6OSSne2aje1yJE=";
  "15.0.7".officialRelease.sha256 = "sha256-wjuZQyXQ/jsmvy6y1aksCcEDXGBjuhpgngF3XQJ/T4s=";
  "16.0.6".officialRelease.sha256 = "sha256-fspqSReX+VD+Nl/Cfq+tDcdPtnQPV1IRopNDfd5VtUs=";
  "17.0.6".officialRelease.sha256 = "sha256-8MEDLLhocshmxoEBRSKlJ/GzJ8nfuzQ8qn0X/vLA+ag=";
  "18.1.8".officialRelease.sha256 = "sha256-iiZKMRo/WxJaBXct9GdAcAT3cz9d9pnAcO1mmR6oPNE=";
  "19.0.0-git".gitRelease = {
    rev = "9b9405621bcc55b74d2177c960c21f62cc95e6fd";
    rev-version = "19.0.0-unstable-2024-06-30";
    sha256 = "sha256-Tlk+caav7e7H6bha8YQyOl+x2iNk9iH7xKpHQkWQyJ4=";
  };
}

# ...

llvmPackages = lib.mapAttrs' (version: args: mkPackage (args // { inherit version; })) versions
```

Another thought is to have a mechanism for overriding/extending it: take `llvmVersions` as a parameter defaulting to empty set. Then do `versions // llvmVersions` -- this would potentially allow users to override LLVM versions by putting `llvmVersions = { ... }` in their overlay. The issue I see with this is of keying the `llvmPackage_{version}` on the major version but having the attribute be a full version.

Reply:
How would making the version list an attr set work with the `officialRelease` or `gitRelease` stuff?

Reply:
That is already accounted for in my suggested code block, `"13.0.1".officialRelease.sha256` expands to `"13.0.1" = { officialRelease = { sha256 = ... }; }` such that the only missing attribute is the version, which we reinject via `mapAttrs'` in `mkPackage (args // { inherit version; }` so that it reads `officialRelease = { version = ..., sha256 = ...}`. Then the last one is `.gitRelease`.

Reply:
Ok, I made it work with an attr set.